### PR TITLE
Bug 1882556: Access git over https for tests

### PR DESCRIPTION
**Pull request by coreydaley-redhat**
Replace git:// urls with https:// urls for all tests

### CODE REVIEW

**Review by adambkaplan — COMMENTED** [review]
I'd like to ensure that we have at least one e2e test that pulls via the `git:` protocol. This test would need to check that we have a cluster-wide proxy configured before running.

**Comment by coreydaley-redhat** [comment]
> I'd like to ensure that we have at least one e2e test that pulls via the `git:` protocol. This test would need to check that we have a cluster-wide proxy configured before running.

So, we would need to fix the proxy (according to the bugzilla) so that git:// urls can be proxied.  Should that be a JIRA then as it is probably a bit more work.  Or are you saying that we need a test that uses a git:// url and make sure that the proxy is NOT configured so that it has a chance to work?

**Comment by adambkaplan** [comment]
> Or are you saying that we need a test that uses a git:// url and make sure that the proxy is NOT configured so that it has a chance to work?

Correct - we should have a test that clones source via `git://`, but when the test starts it checks if a proxy has been configured, skipping if a proxy is present. This should be fairly straightforward (psudo-code here):

```go
func hasProxy() (bool, error) {
    proxy, err := e.ConfigClient.Proxies().Get(ctx, "cluster", metav1.GetOptions{})
    if err != nil && !kerrors.IsNotFound(err) {
        return false, error
    }
    if kerrors.IsNotFound(err) {
        return false, nil
    }
    return len(proxy.Status.HTTPProxy) > 0 || len(proxy.Status.HTTPSProxy) > 0, nil
}
```

**Comment by coreydaley-redhat** [comment]
@adambkaplan ptal

File: test/extended/builds/clone_git_protocol.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = g.Describe("[sig-builds][Feature:Builds] clone repository using git:// protocol", func() {
+	var (
+		oc = exutil.NewCLI("build-pruning")
```
Comment (adambkaplan) [inline test/extended/builds/clone_git_protocol.go:14]: nit - this is the prefix for the generated namespace.
```suggestion
		oc = exutil.NewCLI("build-clone-git-protocol")
```

File: test/extended/builds/clone_git_protocol.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		g.It("should clone using git:// if no proxy is configured", func() {
+			proxyConfigured, err := exutil.HasProxy(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			if proxyConfigured {
+				fmt.Fprintf(g.GinkgoWriter, "Skipping test because proxy is configured")
+				return
+			}
+
+			g.By("creating a new application using the git:// protocol")
+			err = oc.Run("new-app").Args("git://github.com/openshift/ruby-hello-world.git").Execute()
```
Comment (adambkaplan) [inline test/extended/builds/clone_git_protocol.go:47]: Historically we've had issues with `openshift/ruby-hello-world` - the scl example is a bit more stable:

```suggestion
			err = oc.Run("new-app").Args("git://github.com/sclorg/ruby-ex.git").Execute()
```

File: test/extended/util/proxy.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+// HasProxy returns whether or not the currently installed cluster has a proxy configured
+func HasProxy(oc *CLI) (bool, error) {
+	proxy, err := oc.AdminConfigClient().ConfigV1().Proxies().Get(context.Background(), "cluster", metav1.GetOptions{})
+	if err != nil && !kerrors.IsNotFound(err) {
+		return false, err
+	} else if kerrors.IsNotFound(err) {
+		return false, nil
+	}
```
Comment (adambkaplan) [inline test/extended/util/proxy.go:17]: Looks like `kerrors.IsNotFound` returns `false` if the input is nil, so we can simplify a bit:

```suggestion
	if kerrors.IsNotFound(err) {
		return false, nil
	}
	if err != nil {
		return false, err
	}
```

**Review by adambkaplan — COMMENTED** [review]
/approve

Minor suggestions, otherwise looks good.

File: test/extended/util/proxy.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"context"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// HasProxy returns whether or not the currently installed cluster has a proxy configured
+func HasProxy(oc *CLI) (bool, error) {
```
Comment (adambkaplan) [inline test/extended/util/proxy.go:11]: Perhaps need to make it clearer that we're talking about the cluster?

```suggestion
// IsClusterProxyEnabled returns true if the cluster has a global proxy enabled
func IsClusterProxyEnabled(oc *CLI) (bool, error) {
```

File: test/extended/builds/clone_git_protocol.go (hunk 1 above)
Comment (coreydaley-redhat) [reply test/extended/builds/clone_git_protocol.go:14]: updated in next push

File: test/extended/builds/clone_git_protocol.go (hunk 2 above)
Comment (coreydaley-redhat) [reply test/extended/builds/clone_git_protocol.go:47]: updated in next push

File: test/extended/util/proxy.go (hunk 3 above)
Comment (coreydaley-redhat) [reply test/extended/util/proxy.go:17]: updated in next push

File: test/extended/util/proxy.go (hunk 4 above)
Comment (coreydaley-redhat) [reply test/extended/util/proxy.go:11]: updated in next push

**Comment by coreydaley-redhat** [comment]
@adambkaplan all suggestions have been applied, ptal

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Comment by wewang58** [comment]
Using cluster-bot to launch cluster to test it
/lgtm

File: test/extended/builds/clone_git_protocol.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+				exutil.DumpPodLogsStartingWith("", oc)
+			}
+		})
+
+		g.It("should clone using git:// if no proxy is configured", func() {
+			proxyConfigured, err := exutil.IsClusterProxyEnabled(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			if proxyConfigured {
+				fmt.Fprintf(g.GinkgoWriter, "Skipping test because proxy is configured")
+				return
```
Comment (soltysh) [inline test/extended/builds/clone_git_protocol.go:43]: `g.Skip` should be used instead.

Comment (soltysh) [reply test/extended/builds/clone_git_protocol.go:43]: or `e2eskipper.Skipf`.

**Review by soltysh — CHANGES_REQUESTED** [review]
/lgtm cancel

File: test/extended/builds/clone_git_protocol.go (hunk 5 above)
Comment (coreydaley-redhat) [reply test/extended/builds/clone_git_protocol.go]: updated in current push

**Comment by coreydaley-redhat** [comment]
@soltysh Updated per your comment, ptal.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**